### PR TITLE
SI-6077 more conservative TreeMakersToConds for CSE

### PR DESCRIPTION
**Pull request by adriaanm**
we were dropping irrefutable extractor calls (implemented for SI-4691)
in common subcondition elimination, which caused bad code to be generated

clean up TreeMakersToConds significantly in the process

review by @paulp and his crack monkeys

### CODE REVIEW

**Comment by adriaanm** [comment]
another spurious failure (https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/540/consoleText) I can't reproduce locally...

@paulp, could this be related to the partest changes?

```
  [partest] java.lang.AssertionError: assertion failed: scalamap: 7582462360 vs. javamap: 77471315
  [partest] 
  [partest]     at Test$delayedInit$body.apply(t5293-map.scala:78)
  [partest]     at scala.Function0$class.apply$mcV$sp(Function0.scala:40)
  [partest]     at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
  [partest]     at scala.App$$anonfun$main$1.apply(App.scala:61)
  [partest]     at scala.App$$anonfun$main$1.apply(App.scala:61)
  [partest]     at scala.collection.LinearSeqOptimized$class.foreach(LinearSeqOptimized.scala:59)
  [partest]     at scala.collection.immutable.List.foreach(List.scala:77)
  [partest]     at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
  [partest]     at scala.App$class.main(App.scala:61)
  [partest]     at Test$.main(t5293-map.scala:8)
  [partest]     at Test.main(t5293-map.scala)
  [partest]     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  [partest]     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
  [partest]     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
  [partest]     at java.lang.reflect.Method.invoke(Method.java:597)
  [partest]     at scala.tools.nsc.util.ScalaClassLoader$$anonfun$run$1.apply(ScalaClassLoader.scala:71)
  [partest]     at scala.tools.nsc.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
  [partest]     at scala.tools.nsc.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:139)
  [partest]     at scala.tools.nsc.util.ScalaClassLoader$class.run(ScalaClassLoader.scala:71)
  [partest]     at scala.tools.nsc.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:139)
  [partest]     at scala.tools.nsc.CommonRunner$class.run(ObjectRunner.scala:28)
  [partest]     at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:45)
  [partest]     at scala.tools.nsc.CommonRunner$class.runAndCatch(ObjectRunner.scala:35)
  [partest]     at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:45)
  [partest]     at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:70)
  [partest]     at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:92)
  [partest]     at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:101)
  [partest]     at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
  [partest] testing: [...]/files/run/t5293-map.scala                              [FAILED]
```

**Comment by paulp** [comment]
It could be the fault of the partest changes; but I doubt it, except inasmuch as utilizing a machine more efficiently may result in it displaying pathological behavior.  The assert failing here is that a scala for loop shouldn't be more than 10 times slower than a java one; the numbers show it was 100x slower in this case.  We know some of these test machines get bogged down.  I doubt it's more complicated than that.

We need insulation from tests like this - they should exist, but not fail the build.

**Comment by adriaanm** [comment]
I agree. Didn't mean to blame faster partest (but it does seem related). 
We should have a separate category for benchmarks. I think we even have someone working on this.

**Comment by adriaanm** [comment]
ploing! (imagine a somewhat gentler sound than ping and/or bump)

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -2351,11 +2373,15 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
 
       // use the same approximator so we share variables,
       // but need different conditions depending on whether we're conservatively looking for failure or success
-      val reachabilityApproximation = new TreeMakersToConds(prevBinder)
-      val testCasesOk   = reachabilityApproximation.approximateMatch(cases, reachabilityApproximation.constTrue)
-      val testCasesFail = reachabilityApproximation.approximateMatchAgain(cases, reachabilityApproximation.constFalse)
+      // don't rewrite List-like patterns, as List() and Nil need to distinguished for unreachability
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:2376]: You mean, we intentionally fail to notice List() and go quiet about unreachability? What about the other spot where we do the opposite for performance reasons (which as we found the hard way, is not optional) ?

```
/** This is translating uses of List() into Nil.  This is less
 *  than ideal from a consistency standpoint, but it shouldn't be
 *  altered without due caution.
 *  ... this also causes bootstrapping cycles if List_apply is
 *  forced during kind-arity checking, so it is guarded by additional
 *  tests to ensure we're sufficiently far along.
 */
if (args.isEmpty && !forInteractive && fun.symbol.isInitialized && ListModule.hasCompleteInfo && (fun.symbol == List_apply))
  atPos(tree.pos)(gen.mkNil setType restpe)
else
  constfold(treeCopy.Apply(tree, fun, args1) setType ifPatternSkipFormals(restpe))
```

Can this situation be elaborated and cleaned up? I mass-replaced List() with Nil in my branch a while ago after measuring and finding it mattered (not to mdention I much prefer to write Nil to List(), but this made it even more so.)

**Comment by paulp** [comment]
LGTM after accounting for adriaan's limitations.

**Comment by adriaanm** [comment]
.. and https://issues.scala-lang.org/browse/SI-5029, I imagine

**Pull request merged**